cli/command/image/pull_test.go

```diff
@@ -0,0 +1,112 @@
+package image
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/cli/command"
+	"github.com/docker/docker/cli/internal/test"
+	"github.com/docker/docker/pkg/testutil/assert"
+	"github.com/docker/docker/pkg/testutil/golden"
+	"github.com/docker/docker/registry"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"io/ioutil"
+	"testing"
+)
+
+func TestNewPullCommandErrors(t *testing.T) {
+	testCases := []struct {
+		name            string
+		args            []string
+		trustedPullFunc func(ctx context.Context, cli command.Cli, repoInfo *registry.RepositoryInfo, ref reference.Named,
+			authConfig types.AuthConfig, requestPrivilege types.RequestPrivilegeFunc) error
+	}{
+		{
+			name: "wrong-args",
+			args: []string{},
+		},
+		{
+			name: "invalid-name",
+			args: []string{"UPPERCASE_REPO"},
+		},
+		{
+			name: "all-tags-with-tag",
+			args: []string{"--all-tags", "image:tag"},
+		},
+		{
+			name: "pull-error",
+			args: []string{"--disable-content-trust=false", "image:tag"},
+			trustedPullFunc: func(ctx context.Context, cli command.Cli, repoInfo *registry.RepositoryInfo,
+				ref reference.Named, authConfig types.AuthConfig,
+				requestPrivilege types.RequestPrivilegeFunc) error {
+				return errors.Errorf("somethign went wrong")
+			},
+		},
+		{
+			name: "plugin-error",
+			args: []string{"--disable-content-trust=false", "image:tag"},
+			trustedPullFunc: func(ctx context.Context, cli command.Cli, repoInfo *registry.RepositoryInfo,
+				ref reference.Named, authConfig types.AuthConfig,
+				requestPrivilege types.RequestPrivilegeFunc) error {
+				return errors.Errorf("somethign went wrong when fetching 'plugin'")
+			},
+		},
+	}
+	for _, tc := range testCases {
+		buf := new(bytes.Buffer)
+		cmd := NewPullCommand(test.NewFakeCli(&fakeClient{}, buf))
+		cmd.SetOutput(ioutil.Discard)
+		cmd.SetArgs(tc.args)
+		if tc.trustedPullFunc != nil {
+			trustedPull = tc.trustedPullFunc
```

Comment:
Same as above, this is really *magic*, it's really hard to know that it's actually overriding a production code name `trustedPull` by the specified one, and "lower" the test.